core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -217,19 +233,44 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
   def updateCheckpoints(dataDir: File, update: Option[(TopicPartition,Long)]) {
     inLock(lock) {
       val checkpoint = checkpoints(dataDir)
-      val existing = checkpoint.read().filterKeys(logs.keys) ++ update
-      checkpoint.write(existing)
+      if (checkpoint != null) {
+        try {
+          val existing = checkpoint.read().filterKeys(logs.keys) ++ update
+          checkpoint.write(existing)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to access checkpoint file ${checkpoint.f}", e)
+            handleLogDirFailure(checkpoint.f.getParentFile.getAbsolutePath)
+        }
+      }
+    }
+  }
+
+  def handleLogDirFailure(dir: String) {
+    info(s"Stopping cleaning logs in dir $dir")
+    inLock(lock) {
+      checkpoints = checkpoints.filterKeys(_.getAbsolutePath != dir)
```

Comment:
No, that logic is not added. There are two reason I didn't do that. 

One reason is to avoid circular invocation. Currently `LogCleaner.handleLogDirFailure()` and `LogCleanerManager.handleLogDirFailure()` will invoke `LogManager.handleLogDirFailure()`. It seems a bit ugly if we let `LogManager.handleLogDirFailure()` invoke `LogCleanerManager.handleLogDirFailure()`. Ideally the invoke between methods is a directed acyclic graph.

The other reason is that it is not necessary to explicitly remove dir from `LogCleanerManager.checkpoints` as long as `LogManager.handleLogDirFailure()` has already removed all partitions on the offline dir from `LogManager.logs`. Note that all read or write operation on `LogCleanerManager.checkpoints` is triggered by the partition in that log directory, which in turn comes from `LogManager.logs`. Thus if all partitions on a dir is removed from `LogManager.logs`, it is guaranteed that this `dir` will no longer be used as key to access `LogCleanerManager.checkpoints`. Does this address the problem?